### ISSUE TITLE
Random nan values from vmap and grad function

### ISSUE THREAD

**Issue by HaoLi-Math** [google/jax#7069]
I am using Jax 0.2.13 on CPU. When I was training a neural network, I found there is Radom Nans in the output of the gradient of the rural network. The specific problem is as follows.

I found that when I try to feed 500 inputs to the function `vmap_implicit_net_u_t`, there is `nan` in the outputs while if I feed 100 each time, the `nan` disappear. The specific parameters for function `vmap_implicit_net_u_t` can not be uploaded so I put it in [google drive](https://drive.google.com/drive/folders/1ezaNkXYPcXMeXDLeRzPS2FhohNaeEoam?usp=sharing).

The code to reproduce the problem is as follows except the files containing the parameters. I also put it in the [notebook](https://colab.research.google.com/drive/1XHk2adOr8GVWPfo82qsCgCJ0buG49twg?usp=sharing).

```
import jax

from jax.experimental import optimizers
import jax.numpy as jnp
import numpy as np
from jax import grad, jit, vmap, value_and_grad, tree_util
from jax import random
from functools import partial

import matplotlib.pyplot as plt
import scipy.io
from scipy.interpolate import griddata
 
import time
import matplotlib.gridspec as gridspec
from mpl_toolkits.axes_grid1 import make_axes_locatable
import mpmath
import math

import pickle

with open('params_out.pkl', 'rb') as infile:
    params_out = pickle.load(infile)

with open('params_int.pkl', 'rb') as infile:
    params_int = pickle.load(infile)

with open('X_f_train.pkl', 'rb') as infile:
    X_f_train = pickle.load(infile)

nu = 4.84*10**-4
n_x = 200
x_grid = np.linspace(0, 2, num = n_x)
h = 2/(n_x-1)
lb = np.array([0., 0.])
ub = np.array([1., 2.])

# Define Trapezoidal rule
def trap(f, h):
    intgr = h*jnp.sum(f) - 0.5 * h * (f[0] + f[-1])
    # for i in range(1, n-1):
    #     intgr = intgr + h * f[i]
    return intgr

@jit
def net_u(params_out, params_int, t, x):
    t = t.reshape((t.size,))
    x = x.reshape((x.size,))
    X = jnp.concatenate([t,x])
    X = 2.0*(X - lb)/(ub - lb) - 1.0
    num_layers = len(params_out)
    a = params_out[0]
    for l in range(1, num_layers-1):
        W, b = params_out[l]
        X = jnp.tanh(10.0*a*jnp.dot(W, X) + b)
        scaling, shift = params_int[l-1]
        X = scaling * X + shift
    W, b = params_out[-1]
    Y = jnp.dot(W, X) + b
    Y = jnp.squeeze(Y)
    return Y


net_u_t = jit(grad(net_u, 2))
net_u_x = jit(grad(net_u, 3))

vmap_net_u = jit(vmap(net_u, in_axes=(None, None, 0, 0), out_axes=0))
vmap_net_u_x = jit(vmap(net_u_x, in_axes = (None, None, 0, 0), out_axes=0))


@jit
def implicit_net_u(params_out, opt_state_int, t, x):
    for j in range(1):
        # print('interior loop')
        opt_state_int = step_int(j, params_out, opt_state_int, t)
    params_int = get_params_int(opt_state_int)
    return net_u(params_out, params_int, t, x)

vmap_implicit_net_u = jit(vmap(implicit_net_u, in_axes = (None, None, 0, 0), out_axes=0))
vmap_implicit_net_u_t = jit(vmap(grad(implicit_net_u, 2), in_axes = (None, None, 0, 0), out_axes=0))


vmap_net_u_moment = jit(vmap(net_u, in_axes = (None, None, None, 0), out_axes=0))
vmap_net_u_x_moment = jit(vmap(net_u_x, in_axes = (None, None, None, 0), out_axes=0))

def get_invariants_moment(params_out, params_int, nu, t):
    u = vmap_net_u_moment(params_out, params_int, t, x_grid)
    u_x = vmap_net_u_x_moment(params_out, params_int, t, x_grid)
    Invariant1 = trap(u, h) - 0.144598
    Invariant2 = trap(u**2, h) - 0.086759
    Invariant3 = trap(u**3 - 3*nu*u_x**2, h) - 0.046852
    return Invariant1**2 + Invariant2**2 + Invariant3**2

def step_int(i, params_out, opt_state_int, t):
    params_int = get_params_int(opt_state_int)
    g = grad(get_invariants_moment, 1)(params_out, params_int, nu, t)
    return opt_update_int(i, g, opt_state_int)

opt_init, opt_update, get_params = optimizers.adam(step_size=0.0001)
opt_init_int, opt_update_int, get_params_int = optimizers.adam(step_size=0.00001)

opt_state_out = opt_init(params_out)
opt_state_int = opt_init_int(params_int)

ut = vmap_implicit_net_u_t(params_out, opt_state_int, X_f_train[1000:1500, 0], X_f_train[1000:1500, 1])
print(np.isnan(np.min(ut)))

N = 100
for i in range(5):
    ut = vmap_implicit_net_u_t(params_out, opt_state_int, X_f_train[1000+i*N:1000+(i+1)*N, 0], X_f_train[1000+i*N:1000+(i+1)*N, 1])
    print(np.isnan(np.min(ut)))
```

The problem can not be reproduced on GPU. Actually when I try to reproduce it on GPU, there is still a similar problem (large batch fives nan while some batch not), but at different position of X_f_train.

**Comment by apaszke** [comment]
Thanks for the report! This definitely looks sketchy (vmap should not influence numerics), but it will be very difficult for us to nail down the issue given a script of this size. Also, I'm sorry but we cannot run anything that depends on pickled files, as those can execute arbitrary Python code. Isolating a smaller self-contained (incl. no external data) example that exhibits incorrect behaviors will enable us to fix your issue sooner.

**Comment by HaoLi-Math** [comment]
> Thanks for the report! This definitely looks sketchy (vmap should not influence numerics), but it will be very difficult for us to nail down the issue given a script of this size. Also, I'm sorry but we cannot run anything that depends on pickled files, as those can execute arbitrary Python code. Isolating a smaller self-contained (incl. no external data) example that exhibits incorrect behaviors will enable us to fix your issue sooner.

I am sorry for the external data and large size script. But the problem comes up during the training process. I tried a lot but still can not make the example smaller. Is it possible to just copy paste all the parameters here instead of external data? Without that to reproduce my problem I may need to put the whole training process here , which is costly and longer. Sorry for the trouble.

**Comment by hawkinsp** [comment]
Closing since this is pretty old and there's no way to reproduce it, but feel free to reopen (or file a new issue) if the issue still occurs and there is a reproducer we can run.
